src/pudl/analysis/demand_mapping.py

```diff
@@ -1,294 +1,942 @@
-        every intermediate geometry i and source geometry j at index i, j
+        geopandas.GeoDataFrame: GeoDataFrame with all attributes as epas_gdf
+        and one extra attribute with name as the `ID` attribute appended by
+        "_set" substring
 
+        attributes: Adds the `ID`+"_set" as a `constant` attribute and returns the
+        attributes dictionary
     """
-    intersection_matrix = gdf_intersection.pivot_table(
-        values=gdf_intersection_col,
-        index=gdf_intermediate_col,
-        columns=gdf_source_col,
-        fill_value=0
-    )
+    # ID is the index which will help to identify duplicate geometries
+    ID = [k for k, v in attributes.items() if (
+        (k in epas_gdf.columns) and (v == "ID"))][0]
+    gdf_constants = [k for k, v in attributes.items() if (
+        (k in epas_gdf.columns) and (v == "constant"))]
+    gdf_uniforms = [k for k, v in attributes.items() if (
+        (k in epas_gdf.columns) and (v == "uniform"))]
+
+    tqdm_max = min(epas_gdf.shape[0], num_last)
+
+    # Iterating through each of the geometries
+    for index, row in tqdm(epas_gdf[[ID, "geometry"]].iterrows(), total=tqdm_max):
+
+        if index == 0:
+            gdf_disjoint = pd.DataFrame(row).T
+            gdf_disjoint[ID] = gdf_disjoint[ID].apply(lambda x: frozenset([x]))
+            gdf_disjoint = GeoDataFrame(
+                gdf_disjoint, geometry="geometry", crs=epas_gdf.crs)
+            gdf_disjoint_cur_union = unary_union(gdf_disjoint["geometry"])
+
+        # Additional geometries
+        elif index < tqdm_max:
+
+            # Adding difference and intersections of the old geometries
+            # with the new geometry
+            gdf_disjoint["geometry_new_diff"] = gdf_disjoint.difference(
+                row["geometry"])
+            gdf_disjoint["geometry_new_int"] = gdf_disjoint.intersection(
+                row["geometry"])
+            gdf_disjoint = gdf_disjoint.drop("geometry", axis=1)
+
+            # Stacking all the new geometries in one column
+            gdf_disjoint = (gdf_disjoint
+                            .set_index(ID)
+                            .stack()
+                            .reset_index()
+                            .rename(columns={"level_1": "geom_type", 0: "geometry"}))
+
+            # Creating the new ID sets
+            gdf_disjoint[ID] = gdf_disjoint.apply(
+                lambda x: edit_id_set(x, row[ID], ID), axis=1)
+
+            # Adding the new sole geometry's ID and geometry
+            gdf_disjoint = gdf_disjoint.append({
+                ID: frozenset([row[ID]]),
+                "geom_type": "geometry_new_sole",
+                "geometry": row["geometry"].difference(gdf_disjoint_cur_union)
+            }, ignore_index=True)
+
+            # Removing geometries which are not polygons
+            gdf_disjoint["geometry"] = gdf_disjoint["geometry"].apply(
+                lambda x: polygonize_geom(x))
```

Comment:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [Lambda may not be necessary](https://app.codacy.com/manual/zaneselvans/pudl/pullRequest?prid=5664147)